sdk/go/common/resource/plugin/provider-implementers-guide.md

```diff
@@ -0,0 +1,609 @@
+            types of the properties in `inputs` should agree with the types of the
+            resource's input properties as described in its (schema)[#schema]. If `news`
+            contains [`Unknown`s](#unknowns), `inputs` may contain [`Unknown`s](#unknowns).
+- `failures`: any validation failures present in the inputs. These failures should be
+              constrained to type and range mismatches. A failure is a tuple of a
+              [property path](#property-paths) and a failure reason.
+
+#### Diff
+
+The `Diff` method is responsible for calculating the differences between the actual and
+desired state of a resource as represented by its last recorded state and new input
+`Object` as returned from [`Check`](#check) or [`Read`](#read) and the logical
+operation necessary to reconcile the two (i.e. no operation, an `Update, or a `Replace`).
+
+##### Parameters
+
+- `urn`: the [URN](#urns) of the resource.
+- `id`: the [ID](#custom-resources) of the resource.
+- `olds`: the last recorded state `Object` for the resource. This `Object` must have been
+          generated by a call to `Create`, `Read`, or `Update`, and will never contain
+          [`Unknown`s](#unknowns).
+- `news`: the current input `Object` for the resource as returned by [`Check`](#check) or
+          [`Read`](#read). This value may contain [`Unknown`s](#unknowns).
+- `ignoreChanges`: the set of [property paths](#property-paths) to treat as unchanged.
+
+##### Results
+
+- `detailedDiff`: the [detailed diff](#detailed-diffs) between the resource's actual and
+                  desired state.
+- `deleteBeforeReplace`: if true, the resource must be deleted before it is recreated.
+                         This flag is ignored if `detailedDiff` does not indicate that
+                         the resource needs to be replaced.
+- `changes`: an enumeration that indicates whether the provider detected any changes,
+             detected no changes, or does not support detailed diff detection. Providers
+             should return `Some` for this value if there are any entries in
+             `detailedDiff`; otherwise they should return `None` to indicate no
+             difference. If a provider returns `Unknown` for this value, it is the
+             responsibility of the client to determine whether or not differences exist
+             by comparing the resource's last recorded _inputs_ with its current inputs.
+
+In addition, the following properties should be returned for compatibility with older
+clients:
+
+- `replaceKeys`: the list of top-level input property names with changes that require that the
+                 resource be replaced.
+- `stableKeys`: the list of top-level input property names that did not change and
+                top-level output properties that are guaranteed not to change.
+- `changedKeys`: the list of top-level input property names that changed.
```

Comment:
Some of these have different names from the RPC names - which names are we using here?  

(I see `Diffs`, `Stables` and `Replaces` in the RPC proto (and therefore in all our current implementations that implement the raw Go RPC interface))